src/langchain_google_alloydb_pg/async_checkpoint.py

```diff
@@ -0,0 +1,595 @@
+# Copyright 2025 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from typing import Any, AsyncIterator, Iterator, Optional, Sequence, Tuple
+
+from langchain_core.runnables import RunnableConfig
+from langgraph.checkpoint.base import (
+    WRITES_IDX_MAP,
+    BaseCheckpointSaver,
+    ChannelVersions,
+    Checkpoint,
+    CheckpointMetadata,
+    CheckpointTuple,
+    get_checkpoint_id,
+)
+from langgraph.checkpoint.serde.base import SerializerProtocol
+from langgraph.checkpoint.serde.jsonplus import JsonPlusSerializer
+from langgraph.checkpoint.serde.types import TASKS
+from sqlalchemy import text
+from sqlalchemy.ext.asyncio import AsyncEngine
+
+from .engine import CHECKPOINTS_TABLE, AlloyDBEngine
+
+MetadataInput = Optional[dict[str, Any]]
+
+
+class AsyncAlloyDBSaver(BaseCheckpointSaver[str]):
+    """Checkpoint stored in an AlloyDB for PostgreSQL database."""
+
+    __create_key = object()
+
+    jsonplus_serde = JsonPlusSerializer()
+
+    def __init__(
+        self,
+        key: object,
+        pool: AsyncEngine,
+        table_name: str = CHECKPOINTS_TABLE,
+        schema_name: str = "public",
+        serde: Optional[SerializerProtocol] = None,
+    ) -> None:
+        super().__init__(serde=serde)
+        if key != AsyncAlloyDBSaver.__create_key:
+            raise Exception(
+                "only create class through 'create' or 'create_sync' methods"
+            )
+        self.pool = pool
+        self.table_name = table_name
+        self.table_name_writes = f"{table_name}_writes"
+        self.schema_name = schema_name
```

Comment:
Should these be publicly accessible fields

Reply:
We could make them private; however we are currently following the precedent pattern.